plugin/etcd/plugin.go

```diff
@@ -0,0 +1,451 @@
+	if err != nil {
+		fmt.Printf("@R{\u2717 full restore  %s}\n", err)
+		fail = true
+	} else if b {
+		fmt.Printf("@G{\u2713} full restore enabled\n")
+	} else {
+		fmt.Printf("@G{\u2713} full restore disabled\n")
+	}
+
+	s, err = endpoint.StringValueDefault("prefix", "")
+	if err != nil {
+		fmt.Printf("@R{\u2717 prefix  %s}\n", err)
+		fail = true
+	} else if s == "" {
+		fmt.Printf("@R{\u2713 prefix} prefix was not provided so everything will be backed-up/restored\n")
+	} else {
+		fmt.Printf("@G{\u2713 prefix} @C{%s}\n", plugin.Redact(s))
+	}
+
+	if fail {
+		return fmt.Errorf("etcd: invalid configuration")
+	}
+
+	return nil
+}
+
+// Backup ETCD data
+func (p EtcdPlugin) Backup(endpoint plugin.ShieldEndpoint) error {
+	etcd, err := getEtcdConfig(endpoint)
+	if err != nil {
+		return err
+	}
+
+	tlsInfo := transport.TLSInfo{
+		CertFile:      etcd.ClientCertPath,
+		KeyFile:       etcd.ClientKeyPath,
+		TrustedCAFile: etcd.CaCertPath,
+	}
+
+	tlsConfig, err := tlsInfo.ClientConfig()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   []string{etcd.EtcdEndpoints},
+		DialTimeout: 2 * time.Second,
+		Username:    etcd.Username,
+		Password:    etcd.Password,
+		TLS:         tlsConfig,
+	})
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	defer cli.Close()
+	defer cancel()
+
+	resp, err := cli.Get(ctx, etcd.Prefix, clientv3.WithPrefix(), clientv3.WithSort(clientv3.SortByKey, clientv3.SortDescend))
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	for _, ev := range resp.Kvs {
+		fmt.Printf("%s : %s\n", base64.StdEncoding.EncodeToString([]byte(ev.Key)), base64.StdEncoding.EncodeToString([]byte(ev.Value)))
+	}
+
+	return nil
+}
+
+// Restore ETCD data
+func (p EtcdPlugin) Restore(endpoint plugin.ShieldEndpoint) error {
+	reader := bufio.NewReader(os.Stdin)
+
+	etcd, err := getEtcdConfig(endpoint)
+	if err != nil {
+		return err
+	}
+
+	tlsInfo := transport.TLSInfo{
+		CertFile:      etcd.ClientCertPath,
+		KeyFile:       etcd.ClientKeyPath,
+		TrustedCAFile: etcd.CaCertPath,
+	}
+
+	tlsConfig, err := tlsInfo.ClientConfig()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   []string{etcd.EtcdEndpoints},
+		DialTimeout: 2 * time.Second,
+		Username:    etcd.Username,
+		Password:    etcd.Password,
+		TLS:         tlsConfig,
+	})
+
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer cli.Close()
+
+	if etcd.FullOverwrite {
+		if etcd.Prefix == "" {
+			ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+			_, err := cli.Delete(ctx, "", clientv3.WithPrefix())
+			if err != nil {
+				log.Fatal(err)
+			}
+			defer cancel()
+		} else {
+			ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+			_, err := cli.Delete(ctx, etcd.Prefix, clientv3.WithPrefix())
+			if err != nil {
+				log.Fatal(err)
+			}
+			defer cancel()
+		}
```

Comment:
The consequent and alternate clauses of the inner if/else block are identical, given the invariant established by the condition.